pubsub/pubsub.go

```diff
@@ -0,0 +1,272 @@
+	sem chan struct{}
+	q   []*Message
+}
+
+// SubscriptionOptions contains configuration for Subscriptions.
+type SubscriptionOptions struct {
+	// AckDelay tells the max duration to wait before sending the next batch
+	// of acknowledgements back to the server.
+	AckDelay time.Duration
+
+	// AckBatchSize is the maximum number of acks that should be sent to
+	// the server in a batch.
+	AckBatchSize int
+
+	// AckDeadline tells how long the server should wait before assuming a
+	// received message has failed to be processed.
+	AckDeadline time.Duration
+}
+
+// AckDelayDefault is the value for SubscriptionOptions.AckDelay if it is not set.
+const AckDelayDefault = time.Millisecond
+
+// Receive receives and returns the next message from the Subscription's queue,
+// blocking and polling if none are available. This method can be called
+// concurrently from multiple goroutines. The Ack() method of the returned
+// Message has to be called once the message has been processed, to prevent it
+// from being received again.
+func (s *Subscription) Receive(ctx context.Context) (*Message, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-s.sem:
+	}
+	defer func() {
+		s.sem <- struct{}{}
+	}()
+	if len(s.q) == 0 {
+		if err := s.getNextBatch(ctx); err != nil {
+			return nil, err
+		}
+	}
+	m := s.q[0]
+	s.q = s.q[1:]
+	return m, nil
+}
+
+// getNextBatch gets the next batch of messages from the server.
+func (s *Subscription) getNextBatch(ctx context.Context) error {
+	msgs, err := s.driver.ReceiveBatch(ctx)
+	if err != nil {
+		return err
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		if len(msgs) == 0 {
+			return errors.New("subscription driver bug: received empty batch")
+		}
+		s.q = make([]*Message, len(msgs))
+		for i, m := range msgs {
+			s.q[i] = &Message{
+				Body:       m.Body,
+				Attributes: m.Attributes,
+				ackID:      m.AckID,
+				sub:        s,
+			}
+		}
+	}
+	return nil
+}
+
+// Close flushes pending ack sends and disconnects the Subscription.
+func (s *Subscription) Close() error {
+	s.ackBatcher.Flush()
+	return s.driver.Close()
+}
+
+// NewSubscription creates a Subscription from a driver.Subscription and opts to
+// tune sending and receiving of acks and messages. Behind the scenes,
+// NewSubscription spins up a goroutine to gather acks into batches and
+// periodically send them to the server.
+func NewSubscription(ctx context.Context, d driver.Subscription, opts SubscriptionOptions) *Subscription {
+	handler := func(item interface{}) {
+		mecs := item.([]msgErrChan)
+		var ids []driver.AckID
+		for _, mec := range mecs {
+			m := mec.msg
+			id := m.ackID
+			ids = append(ids, id)
+		}
+		err := d.SendAcks(ctx, ids)
+		for _, mec := range mecs {
+			mec.errChan <- err
+		}
+	}
+	ab := bundler.NewBundler(msgErrChan{}, handler)
+	if opts.AckDelay == 0 {
+		ab.DelayThreshold = AckDelayDefault
+	} else {
+		ab.DelayThreshold = opts.AckDelay
+	}
+	if opts.AckBatchSize != 0 {
```

Comment:
What size is this referring to? The number of acks, or the (estimated) byte size of all of the acks? The comment on the Options structs seems to imply the former, but the fact that you've got code computing the size of each (ack, message) implies the latter.